concent_api/middleman/middleman_server.py

```diff
@@ -120,52 +121,87 @@ def _close_middleman(self) -> None:
         self._loop.close()
 
     async def _handle_concent_connection(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter) -> None:
+        tasks = []
+        response_queue = asyncio.Queue(loop=self._loop)
+        self._connection_id = (self._connection_id + 1) % CONNECTION_COUNTER_LIMIT
+        self._response_queue_pool[self._connection_id] = response_queue
         try:
-            frame = await handle_frame_receive_async(reader, settings.CONCENT_PUBLIC_KEY)
-            remote_address = writer.get_extra_info('peername')
-            print("Received %r from %r" % (frame, remote_address))
-            await self._respond_to_user(frame, writer)
-        except (BrokenEscapingInFrameMiddlemanProtocolError, asyncio.LimitOverrunError, MiddlemanProtocolError) as exception:
-            await self._send_immediate_error(writer, exception)
+            request_producer_task = self._loop.create_task(
+                request_producer(self._request_queue, response_queue, reader, self._connection_id)
+            )
+            response_consumer_task = self._loop.create_task(
+                response_consumer(response_queue, writer, self._connection_id)
+            )
+            tasks.append(request_producer_task)
+            tasks.append(response_consumer_task)
+            await request_producer_task  # 1. wait until producer task finishes (Concent will sent no more messages)
+            await asyncio.sleep(PROCESSING_TIMEOUT)  # 2. give some time to process items already put to request queue
+            await response_queue.join()  # 3. wait until items from response queue are processed (sent back to Concent)
+            response_consumer_task.cancel()
+
         except Exception as exception:  # pylint: disable=broad-except
             crash_logger.error(
                 f"Exception occurred: {exception}, Traceback: {traceback.format_exc()}"
             )
             raise
 
+        finally:
+            for task in tasks:  # regardless of exception's occurrence, all unfinished tasks should be cancelled
+                task.cancel()   # if exceptions occurs, producer task might need cancelling as well
+            while response_queue.qsize() > 0:
+                item: ResponseQueueItem = await response_queue.get()  # log all info about not yet retrieved items
+                logger.info(
```

Comment:
Since we dealing with `asyncio`, and "things" are difficult, if it's possible, it would be nice to have a test for this case, when `response_queue.qsize()` is grater then 0.

Reply:
a new class has been added, it contains some tests for that. please check `test_queue_pool`